examples/terminal/src/main.rs

```diff
@@ -12,107 +15,67 @@ fn main() {
-            );
-        }
-        println!();
-    }
+    const TEXT_COLOR: Color = Color::rgb(0xFF, 0xFF, 0xFF);
 
-    // Go back to start
-    print!("{}", cursor::Up(height));
+    // Set up the canvas
+    let width = buffer.size().0;
+    let height = LINE_HEIGHT * buffer.layout_runs().count() as f32;
+    let mut canvas = vec![vec![None; width as usize]; height as usize];
 
-    // Print the buffer
-    let mut last_x = 0;
-    let mut last_y = 0;
-    buffer.draw(&mut swash_cache, text_color, |x, y, w, h, color| {
+    // Draw to the canvas
+    buffer.draw(&mut swash_cache, TEXT_COLOR, |x, y, w, h, color| {
         let a = color.a();
-        if a == 0 || x < 0 || y < 0 || w != 1 || h != 1 {
+        if a == 0 || x < 0 || x >= width as i32 || y < 0 || y >= height as i32 || w != 1 || h != 1 {
             // Ignore alphas of 0, or invalid x, y coordinates, or unimplemented sizes
             return;
         }
 
         // Scale by alpha (mimics blending with black)
-        let scale = |c: u8| cmp::max(0, cmp::min(255, ((c as i32) * (a as i32)) / 255)) as u8;
-
-        // Navigate to x coordinate
-        match x.cmp(&last_x) {
-            Ordering::Greater => {
-                print!("{}", cursor::Right((x - last_x) as u16));
-                last_x = x;
-            }
-            Ordering::Less => {
-                print!("{}", cursor::Left((last_x - x) as u16));
-                last_x = x;
-            }
-            Ordering::Equal => {}
-        }
-
-        // Navigate to y coordinate
-        match y.cmp(&last_y) {
-            Ordering::Greater => {
-                print!("{}", cursor::Down((y - last_y) as u16));
-                last_y = y;
-            }
-            Ordering::Less => {
-                print!("{}", cursor::Up((last_y - y) as u16));
-                last_y = y;
-            }
-            Ordering::Equal => {}
-        }
+        let scale = |c: u8| (c as i32 * a as i32 / 255).clamp(0, 255) as u8;
 
-        // Print a space with the expected color as the background
-        print!(
-            "{} {}",
-            color::Bg(color::Rgb(
-                scale(color.r()),
-                scale(color.g()),
-                scale(color.b()),
-            )),
-            color::Bg(color::Reset),
-        );
-
-        // Printing a space increases x coordinate
-        last_x += 1;
+        let r = scale(color.r());
+        let g = scale(color.g());
+        let b = scale(color.b());
+        canvas[y as usize][x as usize] = Some((r, g, b));
     });
 
-    // Skip over output
-    if last_x > 0 {
-        print!("{}", cursor::Left(last_x as u16));
-    }
-    if (last_y as u16) < height {
-        print!("{}", cursor::Down(height - last_y as u16));
+    // Render the canvas
+    let mut output = String::new();
+
+    for row in canvas {
+        for pixel in row {
+            let (r, g, b) = pixel.unwrap_or((0, 0, 0));
+            write!(&mut output, "{}", "  ".on_truecolor(r, g, b)).ok();
```

Comment:
This string needs to be one space instead of two, otherwise everything renders double sized

Reply:
This was a deliberate choice as I wanted to see square-ish pixels, which is closer to a rendered image. Most monospace fonts people use in the terminal tend to be pretty narrow, approx 1:2.

Various terminal fonts using two spaces:
![two-spaces](https://github.com/pop-os/cosmic-text/assets/38416468/376b9bb8-0edd-4376-a056-25a3323be819)

Various terminal fonts using one space:
![one-space](https://github.com/pop-os/cosmic-text/assets/38416468/0739c531-aa86-4c76-8512-8876b3dbe404)